### ISSUE TITLE
segment fault when using split() in regexp with (??{...})

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7766]
Migrated from [rt.perl.org#33936](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=33936) (status was 'resolved')

Searchable as RT33936$

**Comment by p5pRT** [comment]
### From wangxz00@mails.tsinghua.edu.cn

#### Created by wangxz00@mails.tsinghua.edu.cn


When using split\(\) function in \(??\{\.\.\.\}\)\, the Perl interpreter died\.
This error can be verified by the following program&#8203;:
\#\!/usr/bin/perl
$a="aba";
print "matched\\n" if $a=~/\(??\{"aba"\}\)/;    \# OK
print "matched\\n" if $a=~/\(??\{join\(""\,split\(""\,"aba"\)\)\}\)/; \# Segmentation Fault

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.6:

Configured by ActiveState at Mon Dec 13 09:51:32 2004.

Summary of my perl5 (revision 5 version 8 subversion 6) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -DNO_HASH_SEED -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"F:\Perl\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"F:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    21540 Fix backward-compatibility issues in if.pm
    23565 Wrong MANIFEST.SKIP


@INC for perl v5.8.6:
    F:\Program Files\ActiveState Perl Dev Kit 6.0\lib\
    f:/Perl/lib
    f:/Perl/site/lib
    .


Environment for perl v5.8.6:
    HOME (unset)
    LANG=zh_CN
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=F:\Python24\.;F:\Program Files\ActiveState Perl Dev Kit 6.0\bin\;h:\CTeX\LOCALT~1\ty\bin;h:\CTeX\LOCALT~1\cct\bin;h:\CTeX\texmf\miktex\bin;h:\CTeX\gs\gs8.14\bin;h:\CTeX\WinEdt;E:\WINDOWS\system32\wbem;E:\WINDOWS\system32;f:\perl\bin\;f:\progra~1\aldec\active~1.2\bin;d:\matlab6p5\bin\win32;F:\Program Files\Hummingbird\Connectivity\10.00\Accessories\;;d:\Program Files\UltraEdit;E:\WINDOWS;d:\MinGW\bin;d:\msys\1.0\bin;f:\OrCAD\OrCAD_10.3\tools\fet\bin;f:\OrCAD\OrCAD_10.3\tools\pcb\bin;f:\OrCAD\OrCAD_10.3\tools\specctra\bin;f:\OrCAD\OrCAD_10.3\tools\PSpice\Library;f:\OrCAD\OrCAD_10.3\tools\bin;f:\OrCAD\OrCAD_10.3\tools\Capture;F:\Program Files\ActiveState Komodo 3.0\;f:\PROGRA~1\GNU\WINCVS~1.3\CVSNT;d:\vslick\win;F:\Program Files\Microsoft Visual Studio\Common\Tools\WinNT;F:\Program Files\Microsoft Visual Studio\Common\MSDev98\Bin;F:\Program Files\Microsoft Visual Studio\Common\Tools;F:\Program Files\Microsoft Visual Studio\VC98\bin;f:\Modeltech_6.0\win32;f:\Program Files\STI\bin\pc-win95;;f:\padspwr\specctra\bin
    PERL5LIB=F:\Program Files\ActiveState Perl Dev Kit 6.0\lib\
    PERLDB_OPTS=RemotePort=127.0.0.1:2000
    PERL_BADLANG (unset)
    SHELL (unset)




```
</details>

**Comment by p5pRT** [comment]
### From @rgs

çæå² wrote&#8203;:
> When using split\(\) function in \(??\{\.\.\.\}\)\, the Perl interpreter died\.
> This error can be verified by the following program&#8203;:
> \#\!/usr/bin/perl
> $a="aba";
> print "matched\\n" if $a=~/\(??\{"aba"\}\)/;    \# OK
> print "matched\\n" if $a=~/\(??\{join\(""\,split\(""\,"aba"\)\)\}\)/; \# Segmentation Fault

The regexp engine not being reentrant\, this is not expected to work\.
However\, the segfault seems to occur only with threaded perls\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @hvds

Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;mandrakesoft\.com> wrote&#8203;:
:çæå² wrote&#8203;:
:> When using split\(\) function in \(??\{\.\.\.\}\)\, the Perl interpreter died\.
:> This error can be verified by the following program&#8203;:
:> \#\!/usr/bin/perl
:> $a="aba";
:> print "matched\\n" if $a=~/\(??\{"aba"\}\)/;    \# OK
:> print "matched\\n" if $a=~/\(??\{join\(""\,split\(""\,"aba"\)\)\}\)/; \# Segmentation Fault
:
:The regexp engine not being reentrant\, this is not expected to work\.
:However\, the segfault seems to occur only with threaded perls\.

Just for the record\, as far as I'm concerned the regexp engine is supposed
to be reentrant\, so this is only "not expected to work" until we fix those
bugs\.

Hugo

**Comment by p5pRT** [comment]
### From domo@computer.org

\[A pretty late reply\, this \.\.\.\]

On 2005â02â08\, at 13&#8203;:45\, hv@&#8203;crypt\.org wrote&#8203;:

> Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;mandrakesoft\.com> wrote&#8203;:
> :çæå² wrote&#8203;:
> :> When using split\(\) function in \(??\{\.\.\.\}\)\, the Perl interpreter  
> died\.
> :> This error can be verified by the following program&#8203;:
> :> \#\!/usr/bin/perl
> :> $a="aba";
> :> print "matched\\n" if $a=~/\(??\{"aba"\}\)/;    \# OK
> :> print "matched\\n" if $a=~/\(??\{join\(""\,split\(""\,"aba"\)\)\}\)/; \#  
> Segmentation Fault
> :
> :The regexp engine not being reentrant\, this is not expected to work\.
> :However\, the segfault seems to occur only with threaded perls\.
>
FWIW\, I get a bus error on both threaded and unthreaded bleadperl\,  
although I can make the error go away if the last line is

print "matched\\n" if $a=~/\(??\{join\(m% \*%\,split\("x"\,"aba"\)\)\}\)/;

Having tried other re\-entrant things\, it's just the luck of the draw  
whether and how they blow up\. It's all unsafe\.

> Just for the record\, as far as I'm concerned the regexp engine is  
> supposed
> to be reentrant\, so this is only "not expected to work" until we  
> fix those
> bugs\.

Belatedly\, and until that happy day comes\, I attach a documentation  
patch to describe this limitation\.
\-\- 
Dominic Dunlop

**Comment by p5pRT** [comment]
### From domo@computer.org

<details><summary>perlre_not_re-entrant.patch</summary>

```diff
--- perl-current-reference/pod/perlre.pod	2006-02-01 20:31:03.000000000 +0100
+++ perl-current-working/pod/perlre.pod	2006-04-02 21:12:36.000000000 +0200
@@ -694,6 +694,10 @@
 Better yet, use the carefully constrained evaluation within a Safe
 compartment.  See L<perlsec> for details about both these mechanisms.
 
+Because perl's regex engine is not currently re-entrant, interpolated 
+code may not invoke the regex engine either directly with C<m//> or C<s///>),
+or indirectly with functions such as C<split>.
+
 =item C<(??{ code })>
 X<(??{})>
 X<regex, postponed> X<regexp, postponed> X<regular expression, postponed>
@@ -724,6 +728,10 @@
 	     \)
 	  }x;
 
+Because perl's regex engine is not currently re-entrant, delayed 
+code may not invoke the regex engine either directly with C<m//> or C<s///>),
+or indirectly with functions such as C<split>.
+
 =item C<< (?>pattern) >>
 X<backtrack> X<backtracking>
 

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Dominic Dunlop wrote&#8203;:
> > Just for the record\, as far as I'm concerned the regexp engine is  
> > supposed
> > to be reentrant\, so this is only "not expected to work" until we  
> > fix those
> > bugs\.
> 
> Belatedly\, and until that happy day comes\, I attach a documentation  
> patch to describe this limitation\.

Thanks\, applied as \#27697\.

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by change 91332126cd9c948fe94d99b88f88ed9f528cc0d8\,
which will make its way into 5\.13\.6\, and hence into perl 5\.14\.0\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
